Save Draft and Publish buttons

eustachi0 opened the pull request:
## Description

This PR is related to what it has been discussed on Discord here:
https://discord.com/channels/967097582721572934/1085545085036412999

The new code adds the functionality to disable the Save Draft and the Publish buttons. 

The Save Draft button will be disabled when there are no changes to an existing doc or when creating a new one. If you type the button gets enabled, if you revert/delete what you type, the button will be disabled. The idea is to avoid saving empty and unchanged docs.

The publish button will be disabled when you create a new document, so no empty docs will be published. It will be enabled when there are changes. If changes are reverted/deleted it will be disabled.
The publish button will be disabled when the document has already been published without changes.
The publish button will be enabled when a doc is in draft version. 

I have tested this using large files and most field types, but further tests are needed.

My motivation has been learning more about React and Typescript. I'm a Junior dev, I love React !. I was looking for a CMS to use on a couple of projects I want to develop. I started using Directus and Hygraph (former GraphCMS). I then found PayloadCMS and I fell in love with the DX.

I saw the opportunity to implement this feature within Payload as most CMS do this. I will be happy to keep improving this feature.

Thanks Dan! for suggesting opening a PR. This is in fact my first PR ever :)

- [x] I have read and understand the CONTRIBUTING.md document in this repository

## Type of change

<!-- Please delete options that are not relevant. -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] Existing test suite passes locally with my changes
- [ ] I have made corresponding changes to the documentation

eustachi0 commented on src/admin/components/forms/Form/generateArray.tsx:
```diff
@@ -0,0 +1,32 @@
+import { useWatchForm } from "./context";
```
I forgot to delete this import, I was trying to generate the arrays in this file and then export them, but I got an invalid hook call.

jmikrut commented on src/admin/components/forms/Form/generateArray.tsx:
```diff
@@ -0,0 +1,32 @@
+    };
+
+    // find the key that match 'initialValue' and 'value' (property). Here, only field values will be pushed to the array. For example, when we type in a rich text field for the first time, this action will create an array of children object(s) with the 'current' value(s), but if we backspace or remove what we typed the array of children will remain within getFields() object, so it will always be ...
+    if (obj[property]) {
```
I think the right way to compare is _**when**_ a field changes its value. And then, there will be no need to loop. Every time you write a loop, you should feel like there is a performance downside.

jmikrut reviewed (changes requested):
Hey @eustachi0 — for your first ever PR, you have really jumped in! I am really excited to see how your future as an engineer develops. Not many juniors could get into the code at this level and come up with solutions like this, so bravo.

That said, I have just left a few comments on this PR and I think we really need to do some deep brainstorming together before attempting to implement resolutions to the features you're looking to build. 

My largest concerns are as follows:

1. Performance. We need to absolutely minimize the amount of logic that runs when a field changes, and we need to only look at _**the changed field's value**_ when a given field changes. Right now this PR is running quite a bit of logic on every field, every time any field changes, which will significantly add up and slow down the UI. There is likely an elegant way to do this, and I am willing to help brainstorm how.
2. Simplicity. I think we have all the "pieces" we need to calculate if a doc is indeed able to be saved / published, but we just need to leverage what we have in a more elegant way, rather than writing / running all of this additional logic. 

Off the top of my head, this PR has heavy overlap with the `modified` form state value. If we could simply make the `modified` state value more accurate, we could solve for the majority of what this PR sets out to do.

What if we could write a single effect, that is throttled or debounced to run at a maximum of once every ~250ms? The effect could traverse through each field in state and compare the field `value` to its `initialValue`. If there is a difference, then we know that a field is changed. But if all fields are the same, then we know that the form should not be modified.

This would allow a user to edit a field, but then return it to its initial state, and see that the form is not modified.

What do you think? Are you up for making some changes / taking another shot here? If not, no big deal. My team can take this one at some point. But if you are, great! Just keep in mind that we will be looking for the above two goals (performance, simplicity.)

Looking forward to seeing you around!

eustachi0 commented:
I have done some work on this feature following the requirement of simplicity and performance. As a first proposal after the comments, this is what I have done:

This is an improved version of my previous code.

Simplicity
I think that comparing initial state data with current state data as a whole is simple and straightforward, but it comes with some performance costs.
/form/index.tsx
1. Capture initial state of all fields using data values from getData() 
2. Capture current state (when any field changes) of all fields using data values from getData()
3. Compare the two states above using a function called stateHasChanged()
4. setModified to false only when stateHasChanged() is different to the modified variable value.
		PayloadCMS modified variable current behaviour:
	1. On first load, modified is equal to false
	2. On changes, modified is equal to true
	3. If we revert the changes to the initial state, modified is (still) equal to true

		Using this feature:
	1. On first load, modified is equal to false, stateHasChanged() is equal to false, nothing happens
	2. On changes, modified is equal to true, stateHasChanged() is equal to true, nothing happens
	3. Reverting changes to the initial state, modified is equal to true, at this point stateHasChanged() will be equal to false, here is when modified needs to be changed to false again, so setModified(false) will kick in.

Performance
/form/index.tsx
1. Initial state data
	1. To capture initial state data, I am using a useEffect that runs only when initialState changes, and it has set a timeout of 500ms. Basically, this will only run on first load and when new data is added/saved. The string is stored in a variable using the useRef hook. As getData() is always updating the values in real-time, we ensure here that the data is captured before any changes.
	3. Initial state data comes from an array that has been somehow flatten. This array is a result of a for in loop that runs within getData() to extract the value of all fields. I have improved this for in loop.
			- Why Am I using arrays? Well, let's say I capture the initialState data from getData() into a new object, and then compare this initialState data object to  getData() and forget the extra cost of running for in loops, this would be great, but, i.e. when we create a new document, an input field value is undefined, when we type something it changes, if we then delete what we typed, this input value will be '', so undefined is different to ''; the same happens with a richtext field value, when we create a new document, the value is undefined, when we type something, the value will be a new array with child data, if we delete what we typed, the value will still be the array with child data; same for a checkbox, etc. So, comparing an object created directly from getData() to getData() itself will always be different after changes. I hope this makes sense.
			- Extracting data running a for in loop within the getData() object will ensure that only the value of the fields will be extracted and push to a new array, and then join into a single string. So, a field value that is undefined will be omitted when join runs on the array, if a field is a checkbox and its value is undefined it will be extracted as false, so it will be comparable when it is checked and unchecked. I have to check this using all the field types PayloadCMS provides.
2. Current state data (when there are changes within the fields)
	1. Here is when operations can get expensive, I am using a useEffect that runs when getData() changes, and it has set a timeout of 500ms as well to mitigate the cost of this process. This will run every time a field changes but after 500ms from the last change. The process of creating the current data string is exactly the same as the process of creating the initial state data string as mentioned above.
3. The two string then are compare using a function called stateHasChanged(), this is very simple, if the strings are equal return false, else return true. Then, this value will be used to trigger changing the modified variable as I mentioned previously.

Please, let me know if this approach could work for you.

I have started another branch, to explore another option I have in mind. Basically, I am looking for a way to start with an empty array as the initial state, and then, only push into an array what have changed, so if the array is empty return false, if not return true. I am still exploring this.

More ideas would be appreciated too.
